sql/alps_submitter.go

```diff
@@ -329,30 +381,185 @@ func generateEstimatorCreator(estimator string, attrs []*attribute) (string, err
 	return fmt.Sprintf("tf.estimator.%s(%s)", estimator, strings.Join(cl, ",")), nil
 }
 
-func genALPS(w io.Writer, pr *extendedSelect, fts fieldTypes, db *DB) error {
-	// TODO(uuleon): codegen alps
+func newAlpsFiller(pr *extendedSelect) (*alpsFiller, error) {
+	fcList, fsMap, err := resolveTrainColumns(&pr.columns)
+	if err != nil {
+		return nil, err
+	}
+
+	fssCode := make([]string, 0, len(fsMap))
+	for _, fs := range fsMap {
+		fssCode = append(fssCode, fs.ToString())
+	}
+
+	fcCode, err := generateFeatureColumnCode(fcList)
+	if err != nil {
+		return nil, err
+	}
+
+	attrs, err := resolveTrainAttribute(&pr.attrs)
+	if err != nil {
+		return nil, err
+	}
+
+	odpsAttrs := filter(attrs, "odps")
+	odpsMap := make(map[string]string, len(odpsAttrs))
+	for _, a := range odpsAttrs {
+		odpsMap[a.Name] = a.Value.(string)
+	}
+
+	trainSpecAttrs := filter(attrs, "train_spec")
+	trainMap := make(map[string]string, len(trainSpecAttrs))
+	for _, a := range trainSpecAttrs {
+		trainMap[a.Name] = a.Value.(string)
+	}
+
+	datasetAttrs := filter(attrs, "dataset")
+	datasetMap := make(map[string]string, len(datasetAttrs))
+	for _, a := range datasetAttrs {
+		datasetMap[a.Name] = a.Value.(string)
+	}
+
+	estimatorAttrs := filter(attrs, "estimator")
+
+	estimatorCode, err := generateEstimatorCreator(pr.estimator, estimatorAttrs)
+	if err != nil {
+		return nil, err
+	}
+
+	estimatorCode = strings.Replace(estimatorCode, "\"FC\"", "feature_columns", 1)
+
+	tableName := pr.tables[0]
+
+	fields := make([]string, len(pr.fields))
+	for idx, f := range pr.fields {
+		fields[idx] = fmt.Sprintf("\"%s\"", f)
+	}
+
+	y := &featureSpec{
+		FeatureName: pr.label,
+		IsSparse:    false,
+		Shape:       []int{1},
+		DType:       "int",
+		Delimiter:   ","}
+
+	return &alpsFiller{
+		TableName:            tableName,
+		Fields:               fmt.Sprintf("[%s]", strings.Join(fields, ",")),
+		X:                    fmt.Sprintf("[%s]", strings.Join(fssCode, ",")),
+		Y:                    y.ToString(),
+		OdpsConf:             odpsMap,
+		TrainSpec:            trainMap,
+		DatasetConf:          datasetMap,
+		FeatureColumnCode:    fcCode,
+		EstimatorCreatorCode: estimatorCode}, nil
+}
+
+func genALPS(w io.Writer, pr *extendedSelect) error {
+	r, e := newAlpsFiller(pr)
```

Comment:
In `newAlpsFiller`, the system is referred to as `Alps`; in `genALPS`, it is `ALPS`. Which one should we use?

To my understanding, if the system is named after the name of the mountain, it should be `Alps`; or, if it is a synonym, it should be `ALPS`.

Reply:
You are right, they should be unified into `ALPS`